### ISSUE TITLE
Video/Audio decoder initialisation delay

### ISSUE THREAD

**Issue by JohnRiv** [w3c/me-media-integration-guidelines#4]
_Migrated from https://github.com/w3c/webmediaporting/issues/30:_

If it takes 250ms-500ms to initialise a hardware video/audio decoder, where in the various algorithms would this delay happen?

With [MSE](https://www.w3.org/TR/media-source/), would this be expected to happen in reaction to the initialisation segment being written to the SourceBuffer(s) or could it only happen when the first media data is written?

**Comment by poolec** [comment]
This one is particularly relevant to low latency streaming where you want to be able to kick off decoder initialisation as soon as possible in parallel with choosing an appropriate start point and starting to buffer media data from that point.

If decoder initialisation does not begin when the init segment is written then you end up with either increased latency, increased start-up delay or both, because buffering of media data and decoder initialisation then happens in sequence rather than in parallel.  If any substantial decoder initialisation delay occurs after you have chosen a start point, either latency goes up, or you have to abort the download and choose a new start point, with consequent increased start-up delay.

Things become even worse if decoder initialisation doesn't happen until a substantial quantity of media data has been written to the SourceBuffer(s), e.g. to satisfy an unknown and uncontrollable internal low water mark threshold on an internal buffer.

**Comment by fghilardi** [comment]
**Chrome's Media Stack**  (c/o @sandersdan from https://github.com/w3c/webmediaporting/issues/30#issuecomment-610056448 and https://github.com/w3c/webmediaporting/issues/30#issuecomment-610628219)

> Latency/time of acquisition: we do not acquire codecs at a defined time, and it will vary by platform and video element configuration. MSE may acquire a codec once an initialization segment is available, but may be deferred. New codecs may be acquired at any time to handle codec changes, fallback, or to implement resource optimizations.

> The spec as written assumes infinite resources, and web pages do not typically implement any recovery mechanisms to handle cases where that turns out not to be true. Our workarounds have to be careful to maintain the illusion (basically always eventually emit the expected events) or there will be breakage.

> Note that we are not always successful at maintaining the illusion. For example, we can't seek to the previous keyframe in a realtime stream, and therefore can't use suspend/resume transparently in that case.

> We change our workarounds somewhat routinely, but there may be a core set of principles we can document. (Approximately 'always make forward progress eventually, fake events if you have to'.)

> Comparatively recent changes like MediaSession (media controls in notifications) and the play() promise have helped, in that new players are more flexible when the browser asserts control over video elements. In the past a majority of player implementations would call play() in a loop if we didn't enter a playing state, but that's not a problem we have with modern players.

> I am hopeful that with the right mechanisms now, there will come a future time where we can drop the illusion and just fail playback when resources are constrained. When using software decode or the Page Lifecycle API we are nearly there now.
